src/vm/functions/mod.rs

```diff
@@ -367,3 +369,31 @@ fn special_as_contract(args: &[SymbolicExpression], env: &mut Environment, conte
 
     result
 }
+
+fn special_trait_principal(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    // (trait-principal (..))
+    // arg0 => trait
+    check_argument_count(1, args)?;
+
+    runtime_cost!(cost_functions::TRAIT_PRINCIPAL, env, 0)?;
+
+    let contract_identifier = match &args[0].expr {
+        SymbolicExpressionType::Atom(contract_ref) => {
+            // Dynamic dispatch
+            match context.callable_contracts.get(contract_ref) {
+                Some((ref _contract_identifier, _trait_identifier)) => {
+
+                    env.global_context.database.get_contract(_contract_identifier)
```

Comment:
:nail_care: Same feedback on guard conditions, could this be re-formulated so we avoid the pyramid?

Reply:
looks like it.  Will refactor

Reply:
This code was modeled after the `special_contract_call` function, but you can get the `contract_ref` first, then have a second one afterwards instead of nesting.  Seems to be functionally equivalent to me so I rewrote as suggested.